contracts/treasury/priceOracle/KyberDmmChainLinkPriceOracle.sol

```diff
@@ -0,0 +1,523 @@
+  */
+  function _getExpectedReturnFromToken(
+    IERC20Ext tokenIn,
+    uint256 amountIn,
+    IERC20Ext dest,
+    uint256 destRateEth,
+    uint256 destRateUsd,
+    bool isFromLpToken
+  )
+    internal view
+    returns (uint256 totalReturn)
+  {
+    bool isDestEth = dest == ETH_TOKEN_ADDRESS || dest == IERC20Ext(weth);
+    uint256 rate;
+
+    if (!isFromLpToken) {
+      rate = isDestEth ? getRateOverEth(address(tokenIn)) :
+        _getRateWithDestTokenData(address(tokenIn), destRateEth, destRateUsd);
+      require(rate > 0, '0 aggregator rate');
+      return calculateReturnAmount(amountIn, getDecimals(tokenIn), getDecimals(dest), rate);
+    }
+
+    (IERC20Ext[2] memory tokens, uint256[2] memory amounts) = getExpectedTokensFromLp(
+      address(tokenIn), amountIn
+    );
+
+    uint256 destTokenDecimals = getDecimals(dest);
+
+    // calc equivalent (tokens[0], amounts[0]) -> tokenOuts[0]
+    if (tokens[0] == dest) {
+      rate = PRECISION;
+      totalReturn = totalReturn.add(amounts[0]);
+    } else {
+      rate = isDestEth ? getRateOverEth(address(tokens[0])) :
+        _getRateWithDestTokenData(address(tokens[0]), destRateEth, destRateUsd);
+      require(rate > 0, '0 aggregator rate');
+      totalReturn = totalReturn.add(
+        calculateReturnAmount(amounts[0], getDecimals(tokens[0]), destTokenDecimals, rate)
+      );
+    }
+
+    // calc equivalent (tokens[1], amounts[1]) -> tokenOuts[0]
+    if (tokens[1] == dest) {
+      rate = PRECISION;
+      totalReturn = totalReturn.add(amounts[1]);
+    } else {
+      rate = isDestEth ? getRateOverEth(address(tokens[1])) :
+        _getRateWithDestTokenData(address(tokens[1]), destRateEth, destRateUsd);
+        require(rate > 0, '0 aggregator rate');
+      totalReturn = totalReturn.add(
+        calculateReturnAmount(amounts[1], getDecimals(tokens[1]), destTokenDecimals, rate)
+      );
+    }
+  }
+
+  /**
+  *   @dev Get rate from src token given dest token rates over eth and usd
+  *   It is used to save gas when liquidating multiple tokens or LP tokens
+  */
+  function _getRateWithDestTokenData(
+    address src,
+    uint256 destTokenRateEth,
+    uint256 destTokenRateUsd
+  ) internal view returns (uint256) {
+    if (src == address(ETH_TOKEN_ADDRESS) || src == weth) {
+      if (destTokenRateEth == 0) return 0;
+      return PRECISION.mul(PRECISION) / destTokenRateEth;
+    }
+
+    uint256 rateQuoteEth;
+    uint256 rateQuoteUsd;
+
+    if (destTokenRateEth > 0) {
+      uint256 srcTokenRateEth = getRateOverEth(src);
+      rateQuoteEth = PRECISION.mul(srcTokenRateEth) / destTokenRateEth;
+    }
+
+    if (destTokenRateUsd > 0) {
+      uint256 srcTokenRateUsd = getRateOverUsd(src);
+      rateQuoteUsd = PRECISION.mul(srcTokenRateUsd) / destTokenRateUsd;
+    }
+
+    if (rateQuoteEth == 0) return rateQuoteUsd;
+    if (rateQuoteUsd == 0) return rateQuoteEth;
+    return rateQuoteEth.add(rateQuoteUsd) / 2;
+  }
+
+  function calculateReturnAmount(
```

Comment:
- We can re-use `Utils.calcDstQty`
- add a function to call getSetDecimals for specific tokens to reduce operation gas